### PR TITLE
Check Package::is_installed() before calling methods for installed packages

### PR DESCRIPTION
**Pull request by bradyvercher**
Does something like this take care of the undefined method violations in #66?

### CODE REVIEW

File: src/Provider/PackageArchiver.php [hunk 1]
```diff
@@ -197,7 +197,7 @@ protected function archive_package( string $slug, string $type ): Package {
 		try {
 			$package = $this->packages->first_where( compact( 'slug', 'type' ) );
 
-			if ( null !== $package ) {
+			if ( null !== $package && $package->is_installed() ) {
```
Comment (GaryJones) [inline src/Provider/PackageArchiver.php]: Logically at run time, this should work.

But static analysis will still show that `$package` can be any type of `Package`, including one that hasn't got an `get_installed_release()`. i.e. we're using the human knowledge of having to remember to check for `is_installed()` before relying on `get_installed_release()`. This is something that an interface should be providing.

screenshot 2018-07-11 13 33 38

**Comment by bradyvercher** [comment]
I think the latest commits should clear those issues up.

File: src/Provider/PackageArchiver.php [hunk 2]
```diff
@@ -197,8 +197,8 @@ protected function archive_package( string $slug, string $type ): Package {
 		try {
 			$package = $this->packages->first_where( compact( 'slug', 'type' ) );
 
-			if ( null !== $package ) {
-				$this->release_manager->archive( $package->get_installed_release() );
+			if ( null !== $package && $package->is_installed() ) {
+				$this->release_manager->archive_installed_version( $package );
```
Comment (GaryJones) [inline src/Provider/PackageArchiver.php]: screenshot 2018-07-12 19 39 23

File: src/Archiver.php [hunk 3]
```diff
@@ -47,7 +50,7 @@ public function archive_from_source( Release $release ): string {
 			'tests',
 		], $release );
```
Comment (GaryJones) [inline src/Archiver.php]: screenshot 2018-07-12 19 40 30

File: src/Provider/Upgrade.php [hunk 4]
```diff
@@ -117,8 +117,12 @@ public function maybe_upgrade() {
 	 */
 	protected function cache_packages() {
 		foreach ( $this->repository->all() as $package ) {
+			if ( ! $package->is_installed() ) {
+				continue;
+			}
+
 			try {
-				$this->release_manager->archive( $package->get_installed_release() );
+				$this->release_manager->archive_installed_version( $package );
```
Comment (GaryJones) [inline src/Provider/Upgrade.php]: screenshot 2018-07-12 19 42 53

File: src/ReleaseManager.php [hunk 5]
```diff
@@ -84,11 +84,13 @@ public function archive( Release $release ): Release {
...
 
+		$package    = $release->get_package();
 		$source_url = $release->get_source_url();
+
 		if ( ! empty( $source_url ) ) {
 			$filename = $this->archiver->archive_from_url( $release );
-		} elseif ( $release->get_version() === $release->get_package()->get_installed_version() ) {
-			$filename = $this->archiver->archive_from_source( $release );
+		} elseif ( $package->is_installed_release( $release ) ) {
+			$filename = $this->archiver->archive_from_source( $package, $release->get_version() );
```
Comment (GaryJones) [inline src/ReleaseManager.php]: screenshot 2018-07-12 19 43 34

File: src/ReleaseManager.php [hunk 6]
```diff
@@ -100,6 +102,18 @@ public function archive( Release $release ): Release {
...
+	/**
+	 * Archive the installed version of a package.
+	 *
+	 * @since 0.3.0
+	 *
+	 * @param InstalledPackage $package Installed package.
+	 * @return Release
+	 */
+	public function archive_installed_version( InstalledPackage $package ): Release {
+		return $this->archive( $package->get_intalled_release() );
```
Comment (GaryJones) [inline src/ReleaseManager.php]: screenshot 2018-07-12 19 44 00

File: src/Route/Download.php [hunk 7]
```diff
@@ -126,17 +128,11 @@ protected function send_package( Package $package, string $version ): Response {
...
-		) {
-			$this->release_manager->archive( $release );
-		}
-
-		// Send a 404 if the release isn't available.
-		if ( ! $this->release_manager->exists( $release ) ) {
+		try {
+			// Cache the release if an artifact doesn't already exist.
+			$release = $this->release_manager->archive( $release );
+		} catch ( ExceptionInterface $e ) {
```
Comment (GaryJones) [inline src/Route/Download.php]: screenshot 2018-07-12 19 44 52

**Comment by bradyvercher** [comment]
Figures I'd introduce a couple of bugs trying to fix this. I'll get it worked out when I get a bit of time to take a closer look.

I did download PhpStorm to play around and it looks like just doing a `$package instanceof InstalledPackage` check clears these up.

**Comment by bradyvercher** [comment]
I think the latest commit should clear up these inspection violations. If so, this can be squashed before merging.

**Comment by bradyvercher** [comment]
@GaryJones Did that last commit clean things up?

**Comment by GaryJones** [comment]
Haven't got around to checking this one - tab is still open as a reminder though :-)

**Comment by GaryJones** [comment]
I've still not had a chance to check, but feel free to go ahead and merge - I'll follow up separately if anything is outstanding.

**Comment by bradyvercher** [comment]
Sounds good. I have a couple more things I'd like to look into before hopefully getting a release out soon. Thanks for following up!

**Pull request merged**